### ISSUE TITLE
RustFmt onSave not working in 0.14.0

### ISSUE THREAD

**Issue by conikeec** [saviorisdead/RustyCode#112]
Based on changelog i noticed that you folks addressed this issue.
I noticed that rustfmt or rustfmtonsave still does not work.

My config settings is as follows

Am I missing anything?

// Place your settings in this file to overwrite the default settings
{
    "rust.racerPath": "Users/XXX/.cargo/bin/racer", // Specifies path to Racer binary if it's not in PATH
    "rust.rustLangSrcPath": "/Users/XXX/.local/rustc-nightly/src", // Specifies path to /src directory of local copy of Rust sources
    "rust.rustfmtPath":"Users/XXX/.cargo/bin/rustfmt", // Specifies path to Rustfmt binary if it's not in PATH
    "rust.cargoPath":null, // Specifies path to Cargo binary if it's not in PATH
    "rust.formatOnSave": true, // Turn on/off autoformatting file on save (EXPERIMENTAL)
    "rust.checkOnSave": true, // Turn on/off `cargo check` project on save (EXPERIMENTAL)
    "rust.checkWithClippy": false // Turn on/off `cargo clippy` project on save (EXPERIMENTAL) (cargo-clippy should be installed)	
}

**Comment by Draivin** [comment]
When you run `Format Code` does anything happen? If you run `rustfmt --version` what does it say? And if you run `Users/XXX/.cargo/bin/rustfmt --version`?

**Comment by conikeec** [comment]
When I run the format code extension, I receive the following errors
"Cannot format due to syntax errors"


> rustfmt --version
0.4.1
> which rustfmt
/Users/chetanconikee/.cargo/bin/rustfmt
> /Users/XXXX/.cargo/bin/rustfmt --version
0.4.1

**Comment by Draivin** [comment]
Does the error happen on any file? Even a file containing only a simple main?

If so the problem probably is something OS specific.

**Comment by conikeec** [comment]
I am on OSX, El Capitan 10.11.4 (15E65).
I even tried a clean install (by nuking .code, user setting and finally reinstalling)

The error still persists.

**Comment by junjieliang** [comment]
This is what I found. I'm on OSX 10.11.3.

The `Format Code` command runs the following command:
```
/bin/sh -c /Users/XXX/.cargo/bin/rustfmt --skip-children --write-mode=diff /project/src/main.rs.fmt
```

But according to http://unix.stackexchange.com/questions/144514/add-arguments-to-bash-c the correct command needs to be:
```
/bin/sh -c '/Users/XXX/.cargo/bin/rustfmt $0 $1 $2' --skip-children --write-mode=diff /project/src/main.rs.fmt
```

Unfortunately, now I get an error `error: couldn't read "/project/src/main.rs.fmt": No such file or directory (os error 2)`. What is this `.fmt` file that's supposed to exist, but doesn't?

Hope my (small) investigation can help someone figure out the root problem!

**Comment by Draivin** [comment]
Thanks, that will probably help fixing it, but not having an OSX machine to test it on still makes it pretty inconvenient.

The `.fmt` file is a temporary file created to work around vscode limitations, we create it and run `rustfmt` on it so we can avoid saving a potentially dirty buffer on normal executions of `Format Code`, as soon as `rustfmt` finishes running we read its contents and the file is removed.

**Comment by Draivin** [comment]
@junjieliang @conikeec could you test if the issue is fixed on latest release?

**Comment by junjieliang** [comment]
There are a few things happening here...

1) The original issue is fixed (I'm on RustyCode 0.14.3) - thanks for the prompt patches!

2) There's another issue where `rustfmt` would panic if stdout is not a "real" stdout. I've opened an issue on `rustfmt`: https://github.com/rust-lang-nursery/rustfmt/issues/978, but meanwhile I built a custom binary that removed colored output entirely, to get around this problem. 

3) There's a separate issue where newlines are lost when the diff is applied. As an example, this code:
```
fn test() {
     println!("Hello world!", "This is a long string that", "is greater than 80 characters.");
}
```
gets formatted to:
```
fn test() {    println!("Hello world!",             "This is a long string that",             "is greater than 80 characters.");}
```

Just running `rustfmt` on the command line prints the correct diff with newlines.

It works if I edit the last line of `parseDiff` function in `services/formatService.js` to join newlines with `\n` (`return vscode.TextEdit.replace(range, newLines.join('\n'));`), but I have no idea how that affects other users, especially if I'm the only one encountering problems.

**Comment by Draivin** [comment]
Cannot repro on Windows with rustfmt 0.4.1.

Just curious, but when you run `rustfmt --skip-children --write-mode=diff filename` on the command line, does the line break symbols show up? The current line break behavior is based on them.

Example:
![image](https://cloud.githubusercontent.com/assets/947593/15086070/8c43ddd4-13b4-11e6-8cdf-4d010bf2c749.png)

**Comment by junjieliang** [comment]
I managed to find the bug! It was a problem in `rustfmt` when I tried getting around the panic problem.

I think the onSave functionality is working fine for me now. Thanks for your help!

**Comment by Draivin** [comment]
Glad to help, @saviorisdead I think you can close this now.

**Comment by saviorisdead** [comment]
Yeah, sure.

**Comment by conikeec** [comment]
Seems like no one paid any attention to the error that I called out above.
@junjieliang digressed this thread and everything from there is now lost in translation.

The issue for me is still not resolved and here another issue filed with similar traits
https://github.com/saviorisdead/RustyCode/issues/119

**Comment by saviorisdead** [comment]
@conikeec I am sorry if it would seems that no one noticed your bug reports. They are just very tricky for some reason.
New fix just landed, please try updating `rustfmt` to `0.5` and `RustyCode` to `0.14.4` and let me know if issues resolved.

**Comment by conikeec** [comment]
Thank you very much for the fix @saviorisdead . Works perfectly.

Thanks again
